packages/frontend/src/components/ControlledAutocomplete/AddNewOptionModal.tsx

```diff
@@ -0,0 +1,125 @@
+import { FormEvent, useCallback, useState } from 'react'
+import { useMutation } from '@apollo/client'
+import {
+  FormControl,
+  Modal,
+  ModalBody,
+  ModalCloseButton,
+  ModalContent,
+  ModalHeader,
+  ModalOverlay,
+} from '@chakra-ui/react'
+import { Button, FormLabel, Input } from '@opengovsg/design-system-react'
+import client from 'graphql/client'
+import { CREATE_TABLE } from 'graphql/mutations/tiles/create-table'
+import { GET_DYNAMIC_DATA } from 'graphql/queries/get-dynamic-data'
+
+interface AddNewOptionalModalProps {
+  modalHeader: string
+  onSubmit: (newValue: string) => void
+  onClose: () => void
+}
+
+interface UseCreateNewOptionProps {
+  inputValue: string
+  addNewId?: string
+}
+
+export function useCreateNewOption(setValue: (newValue: string) => void) {
+  const [createTable] = useMutation(CREATE_TABLE)
+  const [isCreatingNewOption, setIsCreatingNewOption] = useState(false)
+  const createNewOption = useCallback(
+    async ({ inputValue, addNewId }: UseCreateNewOptionProps) => {
+      if (!inputValue.trim() || !addNewId) {
+        return
+      }
+      let newValue: string | undefined
+      setIsCreatingNewOption(true)
+      try {
+        switch (addNewId) {
+          case 'tiles-createTileRow-tableId': {
+            const { data } = await createTable({
+              variables: {
+                input: {
+                  name: inputValue.trim(),
+                  isBlank: true,
+                },
+              },
+            })
+            newValue = data?.createTable.id
+            break
+          }
+          default:
+            break
+        }
+        await client.refetchQueries({ include: [GET_DYNAMIC_DATA] })
+        if (newValue) {
+          setValue(newValue)
+        }
+      } finally {
```

Comment:
comment: stacey thinks we're relying too much on the toast error. ideally, we should show errors in the context of where the error occured (e.g. in this case - render as an error `Infobox` in the `Modal` and don't close it)

I think no need for now, but would be good to start moving the UI in that direction

Reply:
👍🏻